numpyro/distributions/mixtures.py

```diff
@@ -0,0 +1,209 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import jax
+import jax.numpy as jnp
+
+from numpyro.distributions import Distribution, constraints
+from numpyro.distributions.discrete import CategoricalLogits, CategoricalProbs
+from numpyro.distributions.util import is_prng_key, validate_sample
+
+
+class MixtureSameFamily(Distribution):
+    """
+    Marginalized Finite Mixture distribution of vectorized components.
+
+    The components being a vectorized distribution implies that all components are from the same family,
+    represented by a single Distribution object.
+
+    :param numpyro.distribution.Distribution mixing_distribution:
+        The mixing distribution to select the components. Needs to be a categorical.
+    :param numpyro.distribution.Distribution component_distribution:
+        Vectorized component distribution.
+
+    As an example:
+
+    .. doctest::
+
+       >>> import jax.numpy as jnp
+       >>> import numpyro.distributions as dist
+       >>> mixing_dist = dist.Categorical(probs=jnp.ones(3) / 3)
+       >>> component_dist = dist.Normal(loc=jnp.zeros(3), scale=jnp.ones(3))
+       >>> mixture = dist.MixtureSameFamily(mixing_dist, component_dist)
+       >>> mixture.batch_shape
+       ()
+       >>> mixture.event_shape
+       ()
+    """
+
+    def __init__(self, mixing_distribution, component_distribution, validate_args=None):
+        self._mixing_distribution = mixing_distribution
+        if not isinstance(
+            self.mixing_distribution, (CategoricalLogits, CategoricalProbs)
+        ):
+            raise ValueError(
+                "The mixing distribution need to be a numpyro.distributions.Categorical. "
+                f"However, it is of type {type(self.mixing_distribution)}"
+            )
+        self._component_distribution = component_distribution
+        if not isinstance(self.component_distribution, Distribution):
+            raise ValueError(
+                "The component distribution need to be a numpyro.distributions.Distribution. "
+                f"However, it is of type {type(self.component_distribution)}"
+            )
+        self._mixture_size = self.mixing_distribution.probs.shape[-1]
+        assert self._component_distribution.batch_shape[-1] == self._mixture_size, (
+            "Component distribution batch shape last dimension "
+            f"(size={self._component_distribution.batch_shape[-1]}) "
+            "needs to correspond to the mixture_size={self._mixture_size}!"
+        )
+        super().__init__(
+            batch_shape=self.mixing_distribution.batch_shape,
+            event_shape=component_distribution.event_shape,
```

Comment:
Not sure what exactly you meant so you might want to double check. I changed it to only use the arguments in the checks (instead of object properties).